pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+		msgc:               make(chan *clusterpb.Part),
+		readyc:             make(chan struct{}),
+		stopc:              stopc,
+		reg:                r,
+		partialStateMergesTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_total",
+			Help: "Number of times we have received a partial state to merge for a key.",
+		}, []string{"key"}),
+		partialStateMergesFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_partial_state_merges_failed_total",
+			Help: "Number of times we have failed to merge a partial state received for a key.",
+		}, []string{"key"}),
+		stateReplicationTotal: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_total",
+			Help: "Number of times we have tried to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+		stateReplicationFailed: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "alertmanager_state_replication_failed_total",
+			Help: "Number of times we have failed to replicate a state to other alertmanagers.",
+		}, []string{"key"}),
+	}
+
+	go s.handleBroadcasts()
+	return s
+}
+
+// AddState adds a new state that will be replicated using the ReplicationFunc. It returns a channel to which the client can broadcast messages of the state to be sent.
+func (s *state) AddState(key string, cs cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return newStateChannel(key, send, s.logger)
+}
+
+// MergePartialState merges a received partial message with an internal state.
+func (s *state) MergePartialState(p *clusterpb.Part) error {
+	s.partialStateMergesTotal.WithLabelValues(p.Key).Inc()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	st, ok := s.states[p.Key]
+	if !ok {
+		level.Error(s.logger).Log("msg", "state key not found while merging", "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return fmt.Errorf("unable to merge state, key '%s' not found while merging", p.Key)
+	}
+
+	if err := st.Merge(p.Data); err != nil {
+		level.Error(s.logger).Log("msg", "failed to merge state", "err", err, "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return errors.Wrap(err, "unable to merge state")
+	}
+
+	return nil
+}
+
+// The clusterWait function uses the position to determine how long should it wait before sending a notification.
+// In the Cortex case, we copy the function directly and use the manager to figure out which position we hold.
+func (s *state) Position() int { return 0 }
```

Comment:
Given the comment, does this method need to be part of `State` interface?

Reply:
It does, I need to update the comment. We actually use the upstream implementation on that gossip-based clustering.

Reply:
Thanks for pointing this out, it made me realise that a) I needed a test as part of this PR and b) that the above implementation is completely wrong.

Reply:
Is this better?